### PR TITLE
Let  `*_apply_dof_transformation` accept lists of cells in Python layer

### LINKED ISSUE #3089: Expose looped versions of dof_transformations
### Describe new/missing feature

To use dof transformations efficiently in Python, it would be good to have an interface where we can send in a contiguous chunk of data, and a set of cell_perm infos, to for instance: `pre_apply_transpose_dof_transformation`, and loop over the data (for instance in nanobind), calling the appropriate function for sub-spans of the original input.

### Suggested user interface

```python3
dof_data = np.arange(num_cells*num_dofs_per_cell, dtype=np.int32)
cell_infos = np.arange(num_cells*num_dofs_per_cell, dtype=np.uint8)
element.pre_apply_transpose_dof_transformation(dof_data, cell_infos, 1)
```

### PR DESCRIPTION
**Pull request by jorgensd**
Makes this function useful from Python when you have a contiguous array relating to more than a single cell with a single permutation.
Resolves #3089.
Tested for multiple cells in: https://github.com/jorgensd/adios4dolfinx/actions/runs/8179344989/job/22365266424?pr=92
while the DOLFINx C++ interface now uses the one cell version (which throws errors in `test_assembly` if not done correctly).

### CODE REVIEW

**Review by garth-wells — COMMENTED** [review]
It would be better to fix this (1) at the 'top' (in the nanobind wrapper) or (2) at the 'bottom' (in Basix). This PR changes it in the 'middle'. The C++ code is more complicated with no benefit to the C++ layer.

**Comment by jorgensd** [comment]
> It would be better to fix this (1) at the 'top' (in the nanobind wrapper) or (2) at the 'bottom' (in Basix). This PR changes it in the 'middle'. The C++ code is more complicated with no benefit to the C++ layer.

@mscroggs what do you think? Should we move this to basix?

File: cpp/dolfinx/fem/FiniteElement.h [hunk 1]
```diff
@@ -468,7 +468,7 @@ class FiniteElement
   /// @param[in,out] data The data to be transformed. This data is flattened
   /// with row-major layout, shape=(num_dofs, block_size)
   /// @param[in] cell_permutation Permutation data for the cell
-  /// @param[in] block_size The block_size of the input data
+  /// @param[in] block_size The block_size of the input data per cell
```
Comment (jorgensd) [inline cpp/dolfinx/fem/FiniteElement.h:471]:
```suggestion
  /// @param[in] block_size The block_size of the input data
```

File: cpp/dolfinx/fem/FiniteElement.h [hunk 2]
```diff
@@ -533,7 +533,7 @@ class FiniteElement
   ///
   /// @param[in,out] data The data to be transformed. This data is flattened
   /// with row-major layout, shape=(num_dofs, block_size)
-  /// @param[in] cell_permutation Permutation data for the cell
+  /// @param[in] cell_permutation Permutation data for the cells
```
Comment (jorgensd) [inline cpp/dolfinx/fem/FiniteElement.h:536]:
```suggestion
  /// @param[in] cell_permutation Permutation data for the cell
```

File: cpp/dolfinx/fem/FiniteElement.h [hunk 3]
```diff
@@ -548,7 +548,7 @@ class FiniteElement
...
```
Comment (jorgensd) [inline cpp/dolfinx/fem/FiniteElement.h:551]:
```suggestion
  /// @param[in] cell_permutation Permutation data for the cell
```

File: cpp/dolfinx/fem/FiniteElement.h [hunk 4]
```diff
@@ -565,7 +565,7 @@ class FiniteElement
   /// @param[in,out] data The data to be transformed. This data is flattened
   /// with row-major layout, shape=(num_dofs, block_size)
   /// @param[in] cell_permutation Permutation data for the cell
-  /// @param[in] block_size The block_size of the input data
+  /// @param[in] block_size The block_size of the input datas
```
Comment (jorgensd) [inline cpp/dolfinx/fem/FiniteElement.h:568]:
```suggestion
  /// @param[in] block_size The block_size of the input data
```

File: cpp/dolfinx/fem/FiniteElement.h [hunk 5]
```diff
@@ -580,7 +580,7 @@ class FiniteElement
   ///
   /// @param[in,out] data The data to be transformed. This data is flattened
   /// with row-major layout, shape=(num_dofs, block_size)
-  /// @param[in] cell_permutation Permutation data for the cell
+  /// @param[in] cell_permutation Permutation data for the cells
```
Comment (jorgensd) [inline cpp/dolfinx/fem/FiniteElement.h:583]:
```suggestion
  /// @param[in] cell_permutation Permutation data for the cell
```

**Comment by jorgensd** [comment]
> It would be better to fix this (1) at the 'top' (in the nanobind wrapper) or (2) at the 'bottom' (in Basix). This PR changes it in the 'middle'. The C++ code is more complicated with no benefit to the C++ layer.

I've moved the fix for this into the nanobind wrapper.
Currently running tests at: https://github.com/jorgensd/adios4dolfinx/actions/runs/8185238997/job/22409020117

**Pull request merged**